### ISSUE TITLE
[P0] double weight_decay — decoupled_weight_decay + optimizer wd stacks

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#13]
## Problem

`UserWarning: decoupled_weight_decay=True but optimizer also has weight_decay=0.1. Double decay! Set optimizer weight_decay=0.`

When `OptimizerConfig.decoupled_weight_decay=True` (the default) and the inner optimizer's param groups carry `weight_decay != 0`, every step applies decay twice:

1. `_apply_decoupled_weight_decay()` (line 2108 of `distrib_optimizer.py`): `param *= (1 - lr * wd)` on FP32 shards — the intended AdamW path. Called at line 2208, **before** `self.optimizer.step()`.
2. `self.optimizer.step()` (line 2211): AdamW / DeepSpeedCPUAdam also applies its own L2 decay because `pg['weight_decay']` was never zeroed.

On the heterogeneous cluster (2×A6000 + H100, PCIe-only) this is especially harmful: `_apply_decoupled_weight_decay` is rank-size-independent by design so all ranks get the same effective decay; but if the inner optimizer's L2 term is also active it re-introduces rank-shard-size dependence (A6000 shards are smaller → proportionally larger gradient-space L2 → different effective wd per rank tier). Convergence diverges silently.

**Reproduction path:**
```python
from deepspeed.core.optimizer.distrib_optimizer import DistributedOptimizer
from deepspeed.core.optimizer import OptimizerConfig
import torch

model = torch.nn.Linear(4, 4)
inner_opt = torch.optim.AdamW(model.parameters(), lr=1e-3, weight_decay=0.1)
cfg = OptimizerConfig(decoupled_weight_decay=True)
opt = DistributedOptimizer(inner_opt, config=cfg, ...)
# UserWarning fires; after opt.step(), param decayed twice
```

## Root Cause

Two independent sub-issues:

1. **`distrib_optimizer.py` lines 1082–1090** (original `__init__` warning block): only emitted a `UserWarning` and called `break` after the first offending param group, leaving all remaining param groups un-zeroed. The intent was to auto-zero, but `break` aborted the loop early.

2. **`dist_opt_adapter.py` lines 82–86** (docstring): written to describe a `wd_mult` mechanism that exists only in `deepspeed/core/optimizer/__init__.py` (a different construction path, lines 44–61). `DistributedOptimizer` never implemented `wd_mult` — it uses `_apply_decoupled_weight_decay()` instead. The docstring was not updated when the actual mechanism was implemented.

### Part 1 — `distrib_optimizer.py` `__init__` (lines 1093–1102) ✅ Merged

Replaced the `break`-and-warn-only block (old lines 1082–1090) with a loop that iterates **all** param groups, emits `logger.info` per offending group, and sets `pg['weight_decay'] = 0.0` unconditionally:

```python
# lines 1093–1102 (current merged state)
if self.config.decoupled_weight_decay:
    for i, pg in enumerate(self.optimizer.param_groups):
        if pg.get('weight_decay', 0.0) != 0.0:
            logger.info(
                f'DistributedOptimizer: param_group[{i}] weight_decay={pg["weight_decay"]} '
                f'zeroed because decoupled_weight_decay=True'
            )
            pg['weight_decay'] = 0.0
```

### Part 2 — `distrib_optimizer.py` `load_state_dict` (lines 1987–1989) ✅ Merged

Added re-zero after `optimizer.load_state_dict()` to handle pre-fix checkpoints that saved `weight_decay != 0`:

```python
# lines 1987–1989 (current merged state)
if self.config.decoupled_weight_decay:
    for pg in self.optimizer.param_groups:
        pg['weight_decay'] = 0.0
```

### Part 3 — `dist_opt_adapter.py` docstring (lines 82–86, 95–96, 101–102) ← **This issue tracks Part 3 only**

**Lines 82–86** — current (incorrect):
```
**Weight decay handling** (fix for issue #13):
When ``wrapped_by_dist_opt=True``, weight_decay is set to 0 in the inner
optimizer because DistributedOptimizer applies weight decay externally via
``param_group['wd_mult']``.  Setting it in both places causes double
application (Megatron convention: inner optimizer weight_decay=0, outer
DistributedOptimizer manages wd_mult per param group).
```

**Lines 82–86** — replacement:
```
**Weight decay handling** (fix for issue #13):
When ``wrapped_by_dist_opt=True``, weight_decay is set to 0 in the inner
optimizer because ``DistributedOptimizer`` applies decoupled weight decay
externally via ``_apply_decoupled_weight_decay()`` (called before
``optimizer.step()``), which reads ``config.weight_decay`` and applies
``param *= (1 - lr * wd)`` directly to FP32 shards.  Setting weight_decay
in both the inner optimizer and via ``_apply_decoupled_weight_decay`` causes
double application.
```

**Lines 95–96** — current (incorrect):
```
    wrapped_by_dist_opt (bool): If True, inner optimizer weight_decay is zeroed
        so DistributedOptimizer can manage wd_mult per param group.
```

**Lines 95–96** — replacement:
```
    wrapped_by_dist_opt (bool): If True, inner optimizer weight_decay is zeroed
        so DistributedOptimizer can apply decoupled decay via _apply_decoupled_weight_decay().
```

**Lines 101–102** — current (incorrect):
```
        # zero wd so outer DistributedOptimizer can manage wd_mult
        inner_wd = 0.0
```

**Lines 101–102** — replacement:
```
        # zero wd; DistributedOptimizer applies decay via _apply_decoupled_weight_decay()
        inner_wd = 0.0
```

## Files

- `deepspeed/core/hetero_bridge/dist_opt_adapter.py`
  - Lines 82–86: fix misleading `wd_mult` docstring → describe `_apply_decoupled_weight_decay()` mechanism (exact replacement above)
  - Lines 95–96: update `wrapped_by_dist_opt` arg description to remove `wd_mult` reference (exact replacement above)
  - Lines 101–102: update inline comment to remove `wd_mult` reference (exact replacement above)

- `deepspeed/core/optimizer/distrib_optimizer.py` — Parts 1 & 2 already merged, no further changes needed.

## Edge Cases

- **`decoupled_weight_decay=True, weight_decay=0.0`**: `config.weight_decay > 0` guard at line 1093 prevents the loop entirely — no-op, correct.
- **`decoupled_weight_decay=False`**: `_apply_decoupled_weight_decay` returns early (line 2125 guard). Inner optimizer's L2 is the only decay mechanism — intended. Auto-correct does NOT fire.
- **`DeepSpeedCPUAdam` inner optimizer**: `adamw_mode=True` applies AdamW-style decay internally reading `pg['weight_decay']` at step time. Zeroing param groups at init and after checkpoint restore is sufficient.
- **Multiple param groups with different `weight_decay` values**: The `__init__` loop at lines 1093–1102 iterates all groups (no `break`) — all are zeroed.
- **Pre-fix checkpoint restore**: `load_state_dict` at lines 1987–1989 re-zeros all param groups after `optimizer.load_state_dict()` regardless of what the saved state contains.

## Acceptance Criteria

- [ ] `dist_opt_adapter.py` lines 82–86, 95–96, 101–102 contain zero occurrences of the string `wd_mult`; the docstring instead references `_apply_decoupled_weight_decay()` as the actual decay mechanism.
- [ ] `grep -n 'wd_mult' deepspeed/core/hetero_bridge/dist_opt_adapter.py` returns empty.
- [ ] `DistributedOptimizer.__init__` with `decoupled_weight_decay=True` and an inner optimizer whose param groups have `weight_decay=0.1`: after construction, all `pg['weight_decay'] == 0.0` (verified via lines 1093–1102).
- [ ] `logger.info` message is emitted once per offending param group (no `break`, no suppression).
- [ ] `_apply_decoupled_weight_decay` (line 2208) fires before `self.optimizer.step()` (line 2211) and applies `param *= (1 - lr * wd)` using `config.weight_decay`.
- [ ] `load_state_dict` with a pre-fix checkpoint (inner optimizer `weight_decay=0.1` in saved state): after loading, all `pg['weight_decay'] == 0.0` (via lines 1987–1989).
- [ ] `HeteroDistOptAdapter.build()` path: no regression — `inner_wd=0.0` from `_build_inner_optimizer` (line 103).
- [ ] `decoupled_weight_decay=False` path: inner optimizer `weight_decay` is not touched.
- [ ] Tests in `tests/test_distrib_optimizer_gaps.py`:
  - Case A: direct `DistributedOptimizer` construction with `AdamW(weight_decay=0.1)` + `decoupled_weight_decay=True` → `assert all(pg['weight_decay'] == 0.0 for pg in opt.optimizer.param_groups)`
  - Case B: same setup with two param groups having different `weight_decay` values → both zeroed, assert each.
  - Case C: `load_state_dict` with a state dict carrying `weight_decay=0.1` → `assert all(pg['weight_decay'] == 0.0 ...)` post-load.
  - Case D: `decoupled_weight_decay=False` → `assert opt.optimizer.param_groups[0]['weight_decay'] == 0.1` (untouched).

## Dependencies

- Blocks: #36 ([EPIC] core/optimizer) — docstring fix (Part 3) required before closing.
- Related: #67 ([AUDIT] core/optimizer + core/pipeline_parallel) — audit should verify no other `DistributedOptimizer` construction sites reference `wd_mult`, and should confirm `load_state_dict` path is covered by tests.
